ddtrace/llmobs/_integrations/crewai.py

```diff
@@ -0,0 +1,189 @@
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+
+from ddtrace.internal.logger import get_logger
+from ddtrace.llmobs import LLMObs
+from ddtrace.llmobs._constants import INPUT_VALUE
+from ddtrace.llmobs._constants import METADATA
+from ddtrace.llmobs._constants import OUTPUT_VALUE
+from ddtrace.llmobs._constants import ROOT_PARENT_ID
+from ddtrace.llmobs._constants import SPAN_KIND
+from ddtrace.llmobs._constants import SPAN_LINKS
+from ddtrace.llmobs._integrations.base import BaseLLMIntegration
+from ddtrace.llmobs._utils import _get_nearest_llmobs_ancestor
+from ddtrace.trace import Pin
+from ddtrace.trace import Span
+
+
+log = get_logger(__name__)
+
+
+class CrewAIIntegration(BaseLLMIntegration):
+    _integration_name = "crewai"
+    _traces_to_task_span_ids = {}  # maps trace_id to list of task span_ids
+    _traces_to_tasks = {}  # maps trace_id to dictionary of task_id to span_id and span_links
+
+    def trace(self, pin: Pin, operation_id: str, submit_to_llmobs: bool = False, **kwargs: Dict[str, Any]) -> Span:
+        if kwargs.get("_ddtrace_ctx"):
+            tracer_ctx, llmobs_ctx = kwargs["_ddtrace_ctx"]
+            pin.tracer.context_provider.activate(tracer_ctx)
+            if self.llmobs_enabled:
+                LLMObs._instance._llmobs_context_provider.activate(llmobs_ctx)
+
+        span = super().trace(pin, operation_id, submit_to_llmobs, **kwargs)
+
+        if kwargs.get("operation") == "crew":
+            self._traces_to_task_span_ids[span.trace_id] = []
+            self._traces_to_tasks[span.trace_id] = {}
+            return span
+        if kwargs.get("operation") == "task":
+            task_id = kwargs.get("instance_id")
+            self._traces_to_task_span_ids[span.trace_id].append(str(span.span_id))
+            task_node = self._traces_to_tasks[span.trace_id].setdefault(task_id, {})
+            task_node["span_id"] = str(span.span_id)
+        return span
+
+    def _get_current_ctx(self, pin):
```

Comment:
I don't love this because we now have to directly import LLMObs in the crewAI integration. Debating whether or not we just rely on signal handling to keep it a relative import only if necessary.